builtin/stash--helper.c

```diff
@@ -0,0 +1,75 @@
+#include "../cache.h"
+#include "../cache-tree.h"
+#include "../diff.h"
+#include "../diffcore.h"
+#include "../gettext.h"
+#include "../git-compat-util.h"
+#include "../revision.h"
+
+static const char builtin_stash__helper_usage[] = {
+	"Usage: git stash--helper --non-patch <i_tree>"
+};
+
+static void diff_callback(struct diff_queue_struct *q,
+					 struct diff_options *options, void *data)
+{
+	int i;
+    fprintf(stderr,"diff_callback\n");
+	for (i = 0; i < q->nr; i++)
+    {
+		struct diff_filepair *p;
+		struct stat st;
+
+        fprintf(stderr,"diff_callback: %d\n", i);
+		p = q->queue[i];
+		add_to_index(&the_index, p->two->path, &st, 0);
+	}
+}
+
+int stash_non_patch(const char *i_tree,	const char *prefix)
+{
+	int result;
+	struct rev_info rev;
+	const char *wt_prefix = NULL;
+	unsigned char sha1[20];
+	const char *me = "git stash--helper";
+
+    fprintf(stderr,"stash_non_patch\n");
+	read_index(&the_index);
+    fprintf(stderr,"read the index\n");
+	init_revisions(&rev, NULL);
+    fprintf(stderr,"initialised rev\n");
+	setup_revisions(0, NULL, &rev, NULL);
+    fprintf(stderr,"setup rev\n");
+	rev.diffopt.output_format |= DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = diff_callback;
+	rev.diffopt.format_callback_data=&result;
+	//diff_setup_done(&rev.diffopt);
+    fprintf(stderr,"prepared diffopts\n");
```

Comment:
>> Have you considered giving the "GNU debugger" a spin?

> Yes, I just need to brush up my gdb-fu big time.

I gave `gdb` a little spin. my current state is this:

````bash
echo 1 > file && git add file && echo unrelated >other-file && git add other-file && git commit -m initial && echo 2 > file && git add file && echo 3 > file
[make-stash-a-builtin d7438b6] initial
 2 files changed, 2 insertions(+)
 create mode 100644 t/file
 create mode 100644 t/other-file
vagrant@vagrant-ubuntu-trusty-64:/vagrant/t$ gdb --args=git stash
gdb: option '--args' doesn't allow an argument
Use `gdb --help' for a complete list of options.
vagrant@vagrant-ubuntu-trusty-64:/vagrant/t$ gdb --args git stash
GNU gdb (Ubuntu 7.7.1-0ubuntu5~14.04.2) 7.7.1
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from git...done.
(gdb) b stash_non_patch
Breakpoint 1 at 0x4966cd: file builtin/stash--helper.c, line 32.
(gdb) r
Starting program: /home/vagrant/bin/git stash
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
stash_non_patch
read the index
initialised rev
setup rev
prepared diffopts
Segmentation fault (core dumped)
Cannot save the current worktree state
[Inferior 1 (process 12289) exited with code 01]
(gdb)
````
The current debugging problem is that i called `gdb --args git stash`, set a break point for `stash_non_patch()`, ran `git stash` from within `gdb` and that launched `git-stash.sh`, which launched another `git` process without the breakpoint and/or without `gdb` attached that went on to the actual call of `stash_non_patch()`. I think I can solve this by modifying `stash_non_patch()` to print it's current pid and then sleep long enough for me to attach `gdb` and set another breakpoint.